### DISCUSSION TITLE
Mixed Property Type

### DISCUSSION THREAD

**Discussion by ghost** [mikro-orm/mikro-orm#2223] [Q&A] [answered]
Hey,
I'm creating a `Setting` entity for my project, containing a key/value pair for configuration. Although I expect `string` to be the most widely used property type, I'd like to support `number`, `string[]` and `number[]` as well. Types might get messy but I have a type-safe solution in mind,
```ts
@Entity()
export class Setting {
	@PrimaryKey()
	key: string;

	@Property()
	value: string;
}
```

Would it be possible to have a mixed property type? I'm not sure if PostgreSQL supports it, since I'm aware SQL is rather very strict with types. What I've been thinking about is converting the data to and storing it as binary, then converting it back when I'm using it. This might be a bit slower initially upon retrieving the value, but since settings will be cached, it'll work fine.

I'm not exactly sure how to go about converting to binary and back. I'd probably have a repository function that uses generics for specifying the return type, though. It should look something along the lines of this pseudocode.
```ts
function get<T extends string | number | string[] | number[]>(key: string) {
	return this.findOne({ key }).deserialize() as T;
}
```
Although, this is not my ideal solution, but I suppose it would work if serialization supports it in MikroORM. If not, do you have any other ideas in mind? Thanks.

**Comment by B4nan** [comment, marked as answer]
Custom types should work fine here.

https://mikro-orm.io/docs/custom-types

Or you could go with a JSON property instead of storing this in a separate table (or have a separate table with JSON column if you don't want to pollute the main table).
